src/libaktualizr/uptane/manifest.cc

```diff
@@ -0,0 +1,75 @@
+#include "manifest.h"
+
+#include "crypto/keymanager.h"
+
+namespace Uptane {
+
+Hash Manifest::installedImageHash() const {
+  // TODO: proper verification of the required fields
+  return Uptane::Hash(Uptane::Hash::Type::kSha256,
+                      (*this)["signed"]["installed_image"]["fileinfo"]["hashes"]["sha256"].asString());
+}
+
+std::string Manifest::signature() const {
+  // TODO: proper verification of the required fields
+  return (*this)["signatures"][0]["sig"].asString();
+}
+
+std::string Manifest::signedBody() const {
+  // TODO: proper verification of the required fields
+  return Utils::jsonToCanonicalStr((*this)["signed"]);
+}
+
+bool Manifest::verifySignature(PublicKey &&pub_key) const {
+  if (!(isMember("signatures") && isMember("signed"))) {
+    LOG_ERROR << "Missing either signature or the signing body/subject: " << *this;
+    return false;
+  }
+
+  return pub_key.VerifySignature(signature(), signedBody());
+}
+
+Manifest ManifestIssuer::sign(const Manifest &manifest, const std::string &report_counter) const {
+  Manifest manifest_to_sign = manifest;
+  if (!report_counter.empty()) {
+    manifest_to_sign["report_counter"] = report_counter;
+  }
+  return key_mngr_->signTuf(manifest_to_sign);
+}
+
+Manifest ManifestIssuer::assembleManifest(const InstalledImageInfo &installed_image_info,
+                                          const Uptane::EcuSerial &ecu_serial) {
+  Json::Value installed_image;
+  installed_image["filepath"] = installed_image_info.name;
+  installed_image["fileinfo"]["length"] = Json::UInt64(installed_image_info.len);
+  installed_image["fileinfo"]["hashes"]["sha256"] = installed_image_info.hash;
+
+  Json::Value unsigned_ecu_version;
+  unsigned_ecu_version["attacks_detected"] = "";
+  unsigned_ecu_version["installed_image"] = installed_image;
+  unsigned_ecu_version["ecu_serial"] = ecu_serial.ToString();
+  unsigned_ecu_version["previous_timeserver_time"] = "1970-01-01T00:00:00Z";
+  unsigned_ecu_version["timeserver_time"] = "1970-01-01T00:00:00Z";
+  return unsigned_ecu_version;
+}
+
+Hash ManifestIssuer::generateVersionHash(const std::string &data) { return Hash::generate(Hash::Type::kSha256, data); }
+
+std::string ManifestIssuer::generateVersionHashStr(const std::string &data) {
+  // think of unifying a hash case,we use both lower and upper cases
```

Comment:
Are you sure about that? Hashes are normally in hex; I don't see the problem.

Reply:
Yeas, I am a number of tests will fail because of hash mismatch. Hashes are represented as strings that include non-numeric characters.

Reply:
Ah, now I see what you mean. I forget we do use `to_upper` in a couple places. I don't recall why. I'd prefer to be consistent as well.